### PR TITLE
Add ports / MCT Model (new PR)

### LINKED ISSUE #3: Add 2D-GRM
In **CADET-Core**, a [2D General Rate Model](https://cadet.github.io/master/interface/unit_operations/2d_general_rate_model.html) is available. Moreover, other 2D units are being developed, such as [MCT](https://github.com/modsim/CADET/pull/95). To support these in CADET-Process, ports need to be added to the `FlowSheet`. 

This is, however, not trivial since flow rates are treated differently in CADET-Process than they are in CADET-Core. While in CADET-Core, all flows between units are always explicitly specified, in CADET-Process they are automatically determined in a preprocessing routine. Here, the concept of `output_state` is introduced which specifies the ratio of all outgoing flows. Consider the following `FlowSheet`:

```mermaid
graph LR
    A[Inlet] -->B[Column]
    B -->|0.5| C[Outlet A]
    B -->|0.5| D[Outlet B]
```

The flow entering the column is split 50/50 to Outlets A and B, respectively.

A 2D unit has multiple inlet ports. Here there are several options:
- **Option 1**: Expose the ports on `FlowSheet` level s.t. users can explicitly define flow from certain upstream units to the corresponding port and then from the outlet port to specific downstream units.
- **Option 2**: Treat the inlet as a single port which distributes all incoming flow according to some internal `input_state` which is recombined at the column outlet before being sent to downstream units according to the unit's `output_state`

Option 1 is more flexible but also more complicated to setup. Also, what's the notion of an `output_state` here? Should there be an output state per port?

Option 2 is simpler but also less flexible.

- [ ]  Handling of nrad in CadetAdapter
- [ ]  Translation between nrad and nports
- [ ]  Adding nrad to create_LWE
- [ ]  Tests

### LINKED ISSUE #70: Add support for parameter multiplexing
For many parameters in CADET-Core, the dimensions can change depending on whether the parameter is assumed to be dependent (or independent) on components, sections, particle types, etc.

E.g., consider the `FILM_DIFFUSION` of the [GRM](https://cadet.github.io/master/interface/unit_operations/general_rate_model.html). Here, the parameter can be specified per
- component
- section
- particle

To avoid ambiguities, a multiplex-parameter is introduced. In this example, `FILM_DIFFUSION_MULTIPLEX` determines whether the parameter is treated as component-, type-, and/or section-independent. This field is optional. When left out, multiplexing behavior is inferred from the length of `FILM_DIFFUSION` (if possible). Valid modes are:
- 0: Component-dependent, type-independent, section-independent; length of  `FILM_DIFFUSION` is `NCOMP`
- 1: Component-dependent, type-independent, section-dependent; length of length of  `FILM_DIFFUSION` is `NCOMP` * `NSEC`; ordering is section-major
- 2: Component-dependent, type-dependent, section-independent; length of  `FILM_DIFFUSION` is `NCOMP` * `NPARTYPE`; ordering is type-major
- 3: Component-dependent, type-dependent, section-dependent; length of  `FILM_DIFFUSION` is `NCOMP` * `NPARTYPE` * `NSEC`; ordering is section-type-major

However, it's not possible to directly mirror this behaviour in CADET-Process since sections are handled differently (using `Events`). Moreover, some features might still be missing, such as multiple particle types.

For this reason, I propose that we add flags (with reasonable defaults) which are then later used to infer the multiplex value generating the config for CADET-Core. This will also allow us to extend the functionality step by step.

In practice, this would look something like this:

```Python
class GRM(UnitBaseClass):
    axial_dispersion = MultiplexedParameter(modes=(axial_dispersion_is_component_dependent, axial_dispersion_is_component_dependent, axial_dispersion_is_component_dependent), default=0)
    axial_dispersion_is_component_dependent = Const(value=True)
    # axial_dispersion_is_section_dependent = Const(value=True) # This one we do not really need since they are always section dependent in CADET-Process
    axial_dispersion_is_particle_type_dependent = Bool(default=False) # Note, multiple particle types are not yet implemented
```

The `MultiplexedParameter` would be responsible of checking the parameter dimension on setting (and getting) by reading the values of the flags.

TODOs:
- [ ] Allow differing flow directions for different channels / radial cells in 2D-models

### PR DESCRIPTION
**Pull request by hannahlanzrath**
(New PR after accidental merge)

This PR adds ports to UnitOperations in CADET process which are required by models such as the 2D-GRM.

For starters, we will test this with the Multi-Channel Transport-Model

To fix #3, we also need to fix #70.

## To do
- [x] Add ports to `add_connection` method
- [x] Add discretization class for MCT
- [x] Add SolutionRecorder for MCT
- [x] Store connections with ports in `FlowSheet`
    - [x] Add tests
    - [x] How to expose discretization n_rad to flow_sheet/ports
- [x] Add check methods
    - [x] Unconnected ports
    - [x] Changes in `discretization` configuration
    - [x] flow rate balance
- [x] Open question: Should ports have individual output states?
- [x] Add parameters to the Name Converter for the MCT in cadetAdapter
- [x] Create a SizedMatrix Type and/or check SizednDArray if it meets our requirements
- [x] Add MCT tests in test_unit_operation
     - [x] add create_MCT
- [x] In unitOperation change flow_direction into array of Switches
- [x] move nchannel from discretization to model parameters 
- [ ] Check dimensionality for all unit operations, see also in solution
     - [ ] Create LWE for all Unit Operations
- [x] Overhaul solution method in simulationResults.py

For new PR?
Decision:
- [ ] Add n_ports as a dimension for solutions
- [ ] Make difference between Radial Coordinates and Channels

For Multiplexing PR?
- [ ] Axial dipersion needs multiplexing to enable dependency on nchannel/state.
- [ ] Flow direction needs multiplexing to enable dependency on nchannel.
- [ ] Update/Add Tests for Multiplexing

## Open question
Should the MCT (and other potential 2D models) inherit from `TubularReactor`? How would this look like.
Are dimensions retained in CADET-Process that are squashed in CADET-Core (see LRMP)? -> Line 740

## Closed question
Should every channel of the MCT have a port by default? --> Yes

### CODE REVIEW

**Comment by daklauss** [comment]
_connections = {
            'inlet': {
                'origins': {
                    None: {                       
                    }
                },
                'destinations': {
                    None: {
                        'column': [None]                   
                    }
                },
            }
        }
        connections = {
            'inlet': {
                'origins': {},
                'destinations': 
                    {
                    'column': None,
                    'mct': ['channel_1'],
                },
            }
        }

**Comment by hannahlanzrath** [comment]
In the current state, solution_bulk **can** handle solution outputs for the MCT with nchannel>1 (and I don't know why). The additional dimension n_ports still needs to be added to solution bulk. When setting up an MCT with nchannel = 1, solution.py drops an error (which I would expect as well in the nchannel>1 case) stemming from the wrong dimensionality. Thus n_ports needs to be added as a dimension in one of the solution classes. 
We could think of making a SolutionClass specifically for Units with ports to handle this?

**Comment by hannahlanzrath** [comment]
In addition to my comment above:

[test_mct_simulation.zip](https://github.com/fau-advanced-separations/CADET-Process/files/14901988/test_mct_simulation.zip)

Not-working example with 1 channel:

```from CADETProcess.processModel import ComponentSystem
from CADETProcess.processModel import Inlet, Outlet, MCT
from CADETProcess.processModel import FlowSheet
from CADETProcess.processModel import Process

from CADETProcess.simulator import Cadet

import numpy as np

component_system = ComponentSystem(1)

mct_flow_sheet = FlowSheet(component_system)

inlet = Inlet(component_system, name='inlet')
mct = MCT(component_system,nchannel=1, name='mct')
outlet = Outlet(component_system, name='outlet1')

mct.solution_recorder.write_solution_bulk = 1

mct_flow_sheet.add_unit(inlet)
mct_flow_sheet.add_unit(mct)
mct_flow_sheet.add_unit(outlet)

mct_flow_sheet.add_connection(inlet, mct, destination_port='channel_0')
mct_flow_sheet.add_connection(mct, outlet, origin_port='channel_0')

process = Process(mct_flow_sheet, 'Tracer_Transport')

inlet.flow_rate = 8e-5

mct.length = 0.3
mct.channel_cross_section_areas = [1,]
mct.axial_dispersion = 0
mct.exchange_matrix = np.array([[0.0]])

process.cycle_time = 120*60

_ = process.add_event('Start', 'flow_sheet.inlet.c', [1], 0)
_ = process.add_event('Stop', 'flow_sheet.inlet.c', [0], 1*60)

process_simulator = Cadet(install_path='') #Add your CADET path pointing to a CADET version with MCT here
process_simulator.check_cadet()
simulation_results = process_simulator.simulate(process)

```

Working example with multiple channels

```
from CADETProcess.processModel import ComponentSystem
from CADETProcess.processModel import Inlet, Outlet, MCT
from CADETProcess.processModel import FlowSheet
from CADETProcess.processModel import Process

from CADETProcess.simulator import Cadet

import numpy as np

component_system = ComponentSystem(1)

mct_flow_sheet = FlowSheet(component_system)

inlet = Inlet(component_system, name='inlet')
mct_3c = MCT(component_system,nchannel=3, name='mct_3c')
mct_2c1 = MCT(component_system,nchannel=2, name='mct_2c1')
mct_2c2 = MCT(component_system,nchannel=2, name='mct_2c2')
outlet1 = Outlet(component_system, name='outlet1')
outlet2 = Outlet(component_system, name='outlet2')

mct_3c.solution_recorder.write_solution_bulk = 1
mct_2c1.solution_recorder.write_solution_bulk = 1
mct_2c2.solution_recorder.write_solution_bulk = 1

mct_flow_sheet.add_unit(inlet)
mct_flow_sheet.add_unit(mct_3c)
mct_flow_sheet.add_unit(mct_2c1)
mct_flow_sheet.add_unit(mct_2c2)
mct_flow_sheet.add_unit(outlet1)
mct_flow_sheet.add_unit(outlet2)

mct_flow_sheet.add_connection(inlet, mct_3c, destination_port='channel_0')
mct_flow_sheet.add_connection(mct_3c, mct_2c1, origin_port='channel_0', destination_port='channel_0')
mct_flow_sheet.add_connection(mct_3c, mct_2c1, origin_port='channel_0', destination_port='channel_1')
mct_flow_sheet.add_connection(mct_3c, mct_2c2, origin_port='channel_1', destination_port='channel_0')
mct_flow_sheet.add_connection(mct_2c1, outlet1, origin_port='channel_0')
mct_flow_sheet.add_connection(mct_2c1, outlet1, origin_port='channel_1')
mct_flow_sheet.add_connection(mct_2c2, outlet2, origin_port='channel_1')

process = Process(mct_flow_sheet, 'Tracer_Transport')

inlet.flow_rate = 8e-5

mct_3c.length = 0.3
mct_3c.channel_cross_section_areas = [1,1,1]
mct_3c.axial_dispersion = 0
mct_3c.exchange_matrix = np.array([[
    [0.0,0.01,0.0],
    [0.02,0.0,0.03],
    [0.0,0.0,0.0]
]])

mct_2c1.length = 0.3
mct_2c1.channel_cross_section_areas = [1,1]
mct_2c1.axial_dispersion = 0
mct_2c1.exchange_matrix = np.array([[
    [0.0,0.0],
    [0.0,0.0],

]])

mct_2c2.length = 0.3
mct_2c2.channel_cross_section_areas = [1,1]
mct_2c2.axial_dispersion = 0
mct_2c2.exchange_matrix = np.array([[
    [0.0,0.0],
    [0.0,0.0],

]])

process.cycle_time = 120*60

_ = process.add_event('Start', 'flow_sheet.inlet.c', [1], 0)
_ = process.add_event('Stop', 'flow_sheet.inlet.c', [0], 1*60)

process_simulator = Cadet(install_path='C:/cadet_from_source/cadet/')
process_simulator.check_cadet()
simulation_results = process_simulator.simulate(process)
```

**Comment by schmoelder** [comment]
Can confirm...

I would have a look at the `CADETAdapter`. Maybe you're inadvertently reading the same solution for inlets and outlets?

**Comment by schmoelder** [comment]
Hey @hannahlanzrath @daklauss Could you please update this PR and let me know what's missing s.t. we can merge?

**Comment by hannahlanzrath** [comment]
In CADET-core nchannel is moved from discretization to model parameters, we will have to account for that in here.

File: CADETProcess/processModel/flowSheet.py [hunk 1]
```diff
@@ -314,12 +365,31 @@ def connections(self):
         add_connection
         remove_connection
 
+        """
+        """
```
Comment (schmoelder) [inline CADETProcess/processModel/flowSheet.py:369]: [Question] Is this code still relevant?

File: CADETProcess/processModel/flowSheet.py [hunk 2]
```diff
@@ -406,9 +524,27 @@ def connection_exists(self, origin, destination):
         destination : UnitBaseClass
             UnitBaseClass where the connection terminates.
 
+        Notes
```
Comment (schmoelder) [inline CADETProcess/processModel/flowSheet.py:527]: [Formatting] Remove if empty

File: CADETProcess/processModel/flowSheet.py [hunk 3]
```diff
@@ -406,9 +524,27 @@ def connection_exists(self, origin, destination):
...
+        -----
+
+
         """
-        if destination in self._connections[origin].destinations \
-                and origin in self._connections[destination].origins:
+        if origin.has_ports and not origin_port:
+            raise CADETProcessError(f'{origin.name} is a Unit Operation with ports, so you need to specify origin_port.')
```
Comment (schmoelder) [inline CADETProcess/processModel/flowSheet.py:533]: [alternative] Consider updating the error message, e.g. "Origin port needs to be specified for unit operations with ports."

Also see below for destination port.

**Review by schmoelder — CHANGES_REQUESTED** [review]
Hey guys, we're almost there, just some formatting / wording before merging.

File: CADETProcess/processModel/unitOperation.py [hunk 4]
```diff
@@ -1161,3 +1202,105 @@ def q(self, q):
...
+
+    Parameters
+    ----------
+    length : UnsignedFloat
+        Length of column.
+    channel_cross_section_areas : List of unsinged floats. Lenght depends on nchannel.
+        Diameter of column.
+    axial_dispersion : UnsignedFloat
+        Dispersion rate of compnents in axial direction.
+    flow_direction : Switch
```
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/processModel/unitOperation.py:1218]: Yes exactly, I added this as a ToDo to the issue you mentioned.

File: CADETProcess/simulationResults.py [hunk 5]
```diff
@@ -142,16 +142,39 @@ def solution(self):
...
-                for i in range(1, self.n_cycles):
-                    solution_complete = np.vstack((
-                        solution_complete, cycles[i].solution_original[1:]
-                    ))
-                solution[unit][sol].time_original = time_complete
-                solution[unit][sol].solution_original = solution_complete
-                solution[unit][sol].reset()
+
+            sol_dep = solutions.copy()
+            not_these = []
```
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/simulationResults.py:147]: @daklauss

File: CADETProcess/simulator/cadetAdapter.py [hunk 6]
```diff
@@ -10,6 +10,7 @@
 import tempfile
 import warnings
 
+import re
```
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/simulator/cadetAdapter.py:13]: I don't think so... @daklauss ?

File: CADETProcess/simulator/cadetAdapter.py [hunk 7]
```diff
@@ -552,85 +553,121 @@ def get_simulation_results(
...
             for unit in process.flow_sheet.units:
+
                 solution[unit.name] = defaultdict(list)
+
+                port_flag = unit.has_ports
+
+                if port_flag:
+                    solution[unit.name]['inlet'] = defaultdict(list)
+                    solution[unit.name]['outlet'] = defaultdict(list)
+#                    solution[unit.name]['bulk'] = defaultdict(list)
```
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/simulator/cadetAdapter.py:564]: @daklauss Do you know why this is here? From my POV I think we can just delete it, but want to make sure.

File: CADETProcess/solution.py [hunk 8]
```diff
@@ -787,9 +787,6 @@ def __init__(
         self.time_original = time
 
         self.axial_coordinates = axial_coordinates
-        # Account for dimension reduction in case of only one cell (e.g. LRMP)
-        if radial_coordinates is not None and len(radial_coordinates) == 1:
-            radial_coordinates = None
```
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/solution.py:792]: Not 100% sure what you mean, I just remember we deleted these 3 lines together after discussing that they're not needed anymore. 
What has that to do with the state ordering in CADET-Core?

File: CADETProcess/processModel/flowSheet.py (hunk 1 above)
Comment (hannahlanzrath) [reply CADETProcess/processModel/flowSheet.py:369]: Passing the ball to @daklauss , was this something we tried and then dropped?

**Comment by hannahlanzrath** [comment]
I will squash and rebase again once I got @daklauss input to the remaining conversations and #154 is merged!!

File: CADETProcess/simulator/cadetAdapter.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (daklauss) [reply CADETProcess/simulator/cadetAdapter.py:13]: I guess there was a moment i tried solving a problem with re but decieded against it, so this line can be removed.

File: CADETProcess/simulator/cadetAdapter.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (daklauss) [reply CADETProcess/simulator/cadetAdapter.py:564]: Can be removed. I initially thought 'bulk' is portdependent.

File: CADETProcess/simulationResults.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daklauss) [reply CADETProcess/simulationResults.py:147]: [suggestion] Maybe `solution_with_ports` ? Units with ports can have both so by iterating over the solution we need to filter them, so we can iterate as if they don't have ports.

[the remark this answers is not in the document]
Comment (schmoelder) [reply CADETProcess/simulationResults.py:147]: I don't fully understand what you mean :see_no_evil:. Maybe we can have a chat about it later.

Btw, the name `sol_dep` could also be a bit clearer.

File: CADETProcess/solution.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/solution.py:792]: From what I see CADET-Process solutions for e.g. CSTR, LRMP don't contain singleton dimensions even without the removed lines, but I am unsure where else that could influence things.

If you're at the stand-up on Monday let's quickly talk about that (and our other open points maybe) so I can add the tests and remove the other lines if needed.

**Comment by hannahlanzrath** [comment]
I rebased and will squash & clean up as a preparation for the merge, once we made the final changes

File: CADETProcess/solution.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (hannahlanzrath) [reply CADETProcess/solution.py:792]: @me See updated TODOs

**Comment by hannahlanzrath** [comment]
I have another small issue to discuss:

As nchannel was a discretization parameter previously, it was handed over to during Unitoperation initialisation 
`mct = MCT(nchannel=3)`

Should we keep that or now handle it like any other model parameters 
`mct.nchannel = 3`

That would also mean we have to update the tests.

File: CADETProcess/solution.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (daklauss) [reply CADETProcess/solution.py:792]: should be okay now

File: CADETProcess/simulationResults.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (daklauss) [reply CADETProcess/simulationResults.py]: done

**Comment by hannahlanzrath** [comment]
Resolved the merge conflicts that were introduced during the last commits and freed the history from diverging branches.

**Comment by hannahlanzrath** [comment]
Cleaned up and added / changed Authorships

**Comment by hannahlanzrath** [comment]
When adding to this branch please now rebase first to avoid merges in the history :)

**Comment by hannahlanzrath** [comment]
@schmoelder Wir bekommen jetzt in der Pipeline 

`ImportError: cannot import name 'VisibleDeprecationWarning' from 'numpy'`

Kann ich da helfen, das zu beheben?

**Comment by schmoelder** [comment]
You're right, this is due to numpy v2.0. For more information, see #159.

**Comment by hannahlanzrath** [comment]
I merged create_LWE into add_ports and rebased. From my side we're good to merge this first version of add_ports support into dev :) Please also have a look at the commits and if they should be even more clustered or not. Thank you <3

**Comment by hannahlanzrath** [comment]
I noticed that ports were not added for parameter sensitivities and added the compatibility. In #166 I want to add tests including sensitivities to ensure we cover that. 

How is our current stand on merging this branch into dev?

**Pull request merged**